### PR TITLE
Fix codegen crashing for `readonly` table value returning

### LINKED ISSUE #40376: [Bug]: Bad type on operand stack error when readonly table is returned
### Description

```ballerina
type Customer record {|
    readonly int id;
    readonly string name;
    int noOfItems;
|};

isolated table<Customer> key(id) customerTable = table [];

public function main() {
    _ = getApiKeywords();
}

isolated function getApiKeywords() returns readonly & table<Customer> key(id) {
    lock {
        return customerTable.cloneReadOnly();
    }
}
```

Stacktrace:
```
[2023-05-09 22:29:46,516] SEVERE {b7a.log.crash} - Bad type on operand stack
Exception Details:
  Location:
    test.main(Lio/ballerina/runtime/internal/scheduling/Strand;)Ljava/lang/Object; @3404: putfield
  Reason:
    Type 'java/lang/Object' (current frame, stack[2]) is not assignable to 'io/ballerina/runtime/internal/values/TableValueImpl'
```

### Steps to Reproduce

_No response_

### Affected Version(s)

v2201.5.0

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by HindujaB**
## Purpose
$subject

Fixes #40376 

## Approach
As for the `readonly` structured type returning, we create frame classes with `Object` fields at code generation. 
But it is not handled properly when storing the frame values with the expected value instances. 
Added necessary casting instructions to fix the issue.

## Samples
```ballerina
type Student record {
    readonly string name;
    int id;
};

function testTableValueReturnFromFunction() {
    table<Student> key(name) & readonly tab1 = getStudentTable();
    table<Student> key(name) tab2 = getStudentTable();
    _ = getStudentTable();
}

function getStudentTable() returns readonly & table<Student> key(name) {
    table<Student> key(name) & readonly studentTable = table [];
    return studentTable;
}
```
## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/methodgen/MethodGen.java [hunk 1]
```diff
@@ -851,24 +852,29 @@ private void generateFrameClassFieldUpdate(List<BIRVariableDcl> localVars, Metho
...
+
+    private void generateFrameClassFieldUpdate(MethodVisitor mv, String frameName, BIRVariableDcl localVar,
+                                               BType bType, int index) {
+        if (TypeTags.isIntegerTypeTag(bType.tag)) {
+            mv.visitVarInsn(LLOAD, index);
+            mv.visitFieldInsn(PUTFIELD, frameName, localVar.jvmVarName, "J");
+        } else if (TypeTags.isStringTypeTag(bType.tag)) {
+            mv.visitVarInsn(ALOAD, index);
+            mv.visitFieldInsn(PUTFIELD, frameName, localVar.jvmVarName,
+                    GET_BSTRING);
```
Comment (Nadeeshan96) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/methodgen/MethodGen.java:867]:
```suggestion
            mv.visitFieldInsn(PUTFIELD, frameName, localVar.jvmVarName, GET_BSTRING);
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/methodgen/MethodGen.java [hunk 2]
```diff
@@ -851,24 +852,29 @@ private void generateFrameClassFieldUpdate(List<BIRVariableDcl> localVars, Metho
...
+        } else if (TypeTags.isXMLTypeTag(bType.tag)) {
+            mv.visitVarInsn(ALOAD, index);
+            mv.visitFieldInsn(PUTFIELD, frameName, localVar.jvmVarName,
+                    GET_XML);
```
Comment (Nadeeshan96) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/methodgen/MethodGen.java:871]:
```suggestion
            mv.visitFieldInsn(PUTFIELD, frameName, localVar.jvmVarName, GET_XML);
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/methodgen/MethodGen.java [hunk 3]
```diff
@@ -851,24 +852,29 @@ private void generateFrameClassFieldUpdate(List<BIRVariableDcl> localVars, Metho
...
+        } else if (TypeTags.REGEXP == bType.tag) {
+            mv.visitVarInsn(ALOAD, index);
+            mv.visitFieldInsn(PUTFIELD, frameName, localVar.jvmVarName,
+                    GET_REGEXP);
```
Comment (Nadeeshan96) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/methodgen/MethodGen.java:875]:
```suggestion
            mv.visitVarInsn(ALOAD, index);
            mv.visitFieldInsn(PUTFIELD, frameName, localVar.jvmVarName, GET_REGEXP);
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/methodgen/MethodGen.java [hunk 4]
```diff
@@ -905,6 +911,7 @@ private void generateFrameClassFieldUpdateByTypeTag(MethodVisitor mv, String fra
                 break;
             case TypeTags.TABLE:
                 mv.visitVarInsn(ALOAD, index);
+                mv.visitTypeInsn(CHECKCAST, TABLE_VALUE_IMPL);
```
Comment (Nadeeshan96) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/methodgen/MethodGen.java:914]: Why do we need the checkcast only for the table case? Shall we check whether there is more correct alternative way to fix this issue? What is the difference between record/map/tuple/array case and the table case?

Comment (warunalakshitha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/methodgen/MethodGen.java:914]: I think we might need to handle for other types as well

Comment (HindujaB) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/methodgen/MethodGen.java:914]: I checked the other value cases. We use API interface variables to store such values (`ArrayValue`, `MapValue` etc.). And these didn't break. 
But for tables, we use `TableValueImpl` values. And they are breaking.

**Review by warunalakshitha — APPROVED** [review]
LGTM

**Review by Nadeeshan96 — APPROVED** [review]
LGTM

**Review by warunalakshitha — APPROVED** [review]

**Pull request merged**